File: apps/passenger-android/src/main/kotlin/lk/mageride/passenger/support/SupportSheets.kt
package lk.mageride.passenger.support

import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.PickVisualMediaRequest
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AttachFile
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ExposedDropdownMenuAnchorType
import androidx.compose.material3.ExposedDropdownMenuBox
import androidx.compose.material3.ExposedDropdownMenuDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import kotlinx.coroutines.launch
import lk.mageride.passenger.R
import lk.mageride.passenger.ui.DateFormat
import lk.mageride.passenger.ui.component.MageRideCta
import lk.mageride.passenger.ui.component.SectionLabel
import lk.mageride.passenger.ui.theme.MageRideTheme
import lk.mageride.shared.data.models.ride.RideHistoryRow
import lk.mageride.shared.data.models.support.TicketDetail
import lk.mageride.shared.data.models.support.TicketEvent
import lk.mageride.shared.util.BusinessCalendar

/**
 * SCR-PA-030's two overlays, hosted together because at most one is ever up.
 *
 * The raise-ticket sheet is the wireframe's own **SCR-PA-030a**. The thread sheet is D2'
 * §SCR-PA-030's *"ticket list/thread"*, which the wireframe draws no separate frame for — so it is a
 * `ModalBottomSheet` over the list the passenger tapped rather than a destination the team-approved
 * baseline has no picture of.
 */
@Composable
internal fun SupportSheets(state: SupportState, model: SupportViewModel) {
    when (state.sheet) {
        SupportSheet.RAISE_TICKET -> RaiseTicketSheet(state = state, model = model)
        SupportSheet.TICKET_THREAD -> TicketThreadSheet(ticket = state.ticket, onDismiss = model::closeSheet)
        null -> Unit
    }
}

/**
 * **SCR-PA-030a · raise a ticket** — an M3 modal bottom sheet (US-16.2).
 *
 * The wireframe's four controls, in its order: **Issue description** (multiline), the **Related
 * trip** dropdown, *"📎 Attach a screenshot"* and **Submit ticket**.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun RaiseTicketSheet(state: SupportState, model: SupportViewModel) {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()

    // The system photo picker: no READ_MEDIA_IMAGES, no storage permission on any API level this
    // app supports, and the grant dies with the pick — which is exactly why `readScreenshot` takes
    // the bytes now rather than parking the Uri in state until Submit.
    val pickScreenshot = rememberLauncherForActivityResult(ActivityResultContracts.PickVisualMedia()) { uri ->
        if (uri != null) {
            scope.launch { model.onScreenshotPicked(readScreenshot(context, uri)) }
        }
    }

    ModalBottomSheet(onDismissRequest = model::closeSheet, sheetState = rememberModalBottomSheetState()) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(MageRideTheme.spacing.md),
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
        ) {
            Text(
                text = stringResource(R.string.support_raise_ticket),
                style = MaterialTheme.typography.titleMedium,
                color = MaterialTheme.colorScheme.onSurface,
            )

            OutlinedTextField(
                value = state.description,
                onValueChange = model::onDescriptionChange,
                modifier = Modifier.fillMaxWidth(),
                label = { Text(text = stringResource(R.string.support_issue_label)) },
                placeholder = { Text(text = stringResource(R.string.support_issue_hint)) },
                minLines = MIN_DESCRIPTION_LINES,
                singleLine = false,
            )

            SectionLabel(text = stringResource(R.string.support_related_trip))
            TripDropdown(state = state, onSelect = model::onTripSelected)

            OutlinedButton(onClick = { pickScreenshot.launch(imageOnly()) }, modifier = Modifier.fillMaxWidth()) {
                Icon(imageVector = Icons.Filled.AttachFile, contentDescription = null)
                Text(
                    text = stringResource(
                        if (state.screenshot == null) R.string.support_attach else R.string.support_attached,
                    ),
                    modifier = Modifier.padding(start = MageRideTheme.spacing.xxs),
                )
            }

            MageRideCta(
                label = stringResource(R.string.support_submit_ticket),
                onClick = model::submit,
                enabled = state.canSubmit,
                loading = state.submitting,
            )
        }
    }
}

/**
 * The wireframe's *"PAX-90431-0617 · Nugegoda → Galle Face ▾"*.
 *
 * The id in the wireframe is a passenger-and-date composite this platform does not mint (the same
 * finding C083 recorded about `PAX-90431`), so the row is the **day and the route** — which is what
 * a passenger recognises a trip by, and what the support agent will search on. Optional: a complaint
 * about the app itself is not about a trip at all.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun TripDropdown(state: SupportState, onSelect: (String?) -> Unit, modifier: Modifier = Modifier) {
    var expanded by remember { mutableStateOf(false) }
    val selected = state.trips.firstOrNull { it.rideId == state.tripId }

    ExposedDropdownMenuBox(
        expanded = expanded,
        onExpandedChange = { expanded = !expanded },
        modifier = modifier.fillMaxWidth(),
    ) {
        OutlinedTextField(
            value = selected?.let { tripLabel(it) } ?: stringResource(R.string.support_trip_none),
            onValueChange = {},
            modifier = Modifier
                .fillMaxWidth()
                // `ExposedDropdownMenuAnchorType` is the Material3 1.4 spelling; the argument-less
                // `menuAnchor()` is deprecated there.
                .menuAnchor(ExposedDropdownMenuAnchorType.PrimaryNotEditable),
            readOnly = true,
            singleLine = true,
            trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = expanded) },
        )
        ExposedDropdownMenu(expanded = expanded, onDismissRequest = { expanded = false }) {
            DropdownMenuItem(
                text = { Text(text = stringResource(R.string.support_trip_none)) },
                onClick = {
                    onSelect(null)
                    expanded = false
                },
            )
            state.trips.forEach { trip ->
                DropdownMenuItem(
                    text = { Text(text = tripLabel(trip)) },
                    onClick = {
                        onSelect(trip.rideId)
                        expanded = false
                    },
                )
            }
        }
    }
}

/** One ticket and its whole conversation (US-16.2). */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun TicketThreadSheet(ticket: TicketDetail?, onDismiss: () -> Unit) {
    ModalBottomSheet(onDismissRequest = onDismiss, sheetState = rememberModalBottomSheetState()) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(MageRideTheme.spacing.md)
                .verticalScroll(rememberScrollState()),
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
        ) {
            if (ticket == null) {
                Text(
                    text = stringResource(R.string.support_loading),
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                return@Column
            }

            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
            ) {
                Text(
                    text = categoryLabel(ticket.category),
                    modifier = Modifier.weight(1f),
                    style = MaterialTheme.typography.titleMedium,
                    color = MaterialTheme.colorScheme.onSurface,
                )
                StatusPill(status = ticket.status)
            }

            Text(
                text = ticket.description,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurface,
            )

            // Oldest first, as the contract sends it: a thread read bottom-up is a thread nobody
            // reads. `assigned` entries are skipped — who is handling a complaint is not the
            // complainant's to see, which is the contract's own rule rather than this screen's.
            ticket.thread.forEach { event -> ThreadEntry(event = event) }
        }
    }
}

/** One `TicketEvent` — what happened, and the agent's words when there are any. */
@Composable
private fun ThreadEntry(event: TicketEvent, modifier: Modifier = Modifier) {
    val label = SupportLabels.event(event.kind) ?: return

    Column(modifier = modifier.fillMaxWidth()) {
        Text(
            text = stringResource(label),
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        event.body?.takeIf(String::isNotBlank)?.let { body ->
            Text(
                text = body,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurface,
            )
        }
    }
}

/**
 * *"12 Jun · Nugegoda → Galle Face"*.
 *
 * The date is read in **Colombo** through `BusinessCalendar.ZONE`, not in the handset's zone
 * (D-38): a passenger naming yesterday's trip to support must name the same day support sees.
 */
@Composable
private fun tripLabel(trip: RideHistoryRow): String {
    val route = listOfNotNull(trip.pickup?.address, trip.dropoff?.address)
        .joinToString(ARROW)
        .ifBlank { stringResource(R.string.support_trip_unnamed) }

    return "${DateFormat.dayMonth(BusinessCalendar.businessDate(trip.completedAt))} $SEPARATOR $route"
}

private fun imageOnly() = PickVisualMediaRequest(ActivityResultContracts.PickVisualMedia.ImageOnly)

/** The wireframe's `min-height:72px` description box, as lines. */
private const val MIN_DESCRIPTION_LINES = 3

private const val ARROW = " → "
private const val SEPARATOR = "·"

File: apps/passenger-ios/PassengerApp/Support/SupportSheets.swift
import MageRideShared
import PhotosUI
import SwiftUI

/// SCR-PI-030's two overlays, hosted together because at most one is ever up.
///
/// The raise-ticket sheet is the wireframe's own **SCR-PI-030a**, and its states line fixes the
/// presentation: *"iOS `.sheet` (detent `.medium`)"*. The thread sheet is D2' §SCR-PI-030's *"ticket
/// list/thread"*, which the wireframe draws no separate frame for — so it is a `.sheet` over the
/// list the passenger tapped rather than a destination the team-approved baseline has no picture of.
/// `apps/passenger-android` makes the same call with a `ModalBottomSheet`.
@MainActor
struct SupportSheets: View {

    let sheet: SupportSheet
    @ObservedObject var model: SupportModel

    var body: some View {
        switch sheet {
        case .raiseTicket:
            RaiseTicketSheet(model: model)
        case .ticketThread:
            TicketThreadSheet(ticket: model.state.ticket)
        }
    }
}

/// **SCR-PI-030a · raise a ticket** — a `.sheet` at detent `.medium` (US-16.2).
///
/// The wireframe's four controls, in its order: **Issue description** (the `min-height:72px` box),
/// the **Related trip** field with its `▾`, *"📎 Attach a screenshot"* and **Submit ticket**.
///
/// `.large` is offered beside `.medium` for ``LabelledTextField``'s reason on SCR-PI-026a: a
/// three-line description at an accessibility content size does not fit half a screen, and a second
/// detent is what stops the CTA sitting under the keyboard.
@MainActor
private struct RaiseTicketSheet: View {

    @ObservedObject var model: SupportModel

    /// The system photo picker: **no `NSPhotoLibraryUsageDescription`**, because `PhotosPicker` is
    /// PHPicker and runs out of process, granting access to the one image the passenger chose. The
    /// same contract C100's transfer slip uses, and the reason this app's `Info.plist` carries no
    /// photo-library purpose string.
    @State private var picked: PhotosPickerItem?

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: MageRideSpacing.sm) {
                Text(key: "support_raise_ticket")
                    .mageFont(.title)
                    .foregroundStyle(MageRideColor.onSurface)

                MultilineTextField(
                    labelKey: "support_issue_label",
                    placeholderKey: "support_issue_hint",
                    value: Binding(get: { model.state.description }, set: model.onDescriptionChange)
                )

                SectionLabel(key: "support_related_trip")
                tripPicker

                attachButton

                Button { Task { await model.submit() } } label: {
                    Text(key: "support_submit_ticket")
                }
                .buttonStyle(.mageCta(loading: model.state.isSubmitting))
                .disabled(!model.state.canSubmit)

                if let errorKey = model.state.errorKey {
                    FormErrorText(messageKey: errorKey)
                }
            }
            .padding(MageRideSpacing.md)
        }
        .background(MageRideColor.background)
        .presentationDetents([.medium, .large])
        .onChange(of: picked) { item in
            Task { await load(item) }
        }
    }

    /// The wireframe's *"PAX-90431-0617 · Nugegoda → Galle Face ▾"*.
    ///
    /// **Δ iOS — a `Picker`, which is what the cell's own states line spells out** (*"past **Trip
    /// ID** `Picker`"*) where the Android twin builds an `ExposedDropdownMenuBox`. `.menu` style, so
    /// it reads as the wireframe's `▾` field rather than as a wheel; what it shows is the day and the
    /// route, for ``SupportLabels/trip(_:)``'s reason.
    ///
    /// Optional, and *"No trip"* is a real option rather than a placeholder: a complaint about the
    /// app itself is not about a trip at all.
    private var tripPicker: some View {
        Picker(
            selection: Binding(get: { model.state.tripId }, set: model.onTripSelected),
            label: Text(key: "support_related_trip")
        ) {
            Text(key: "support_trip_none").tag(String?.none)
            ForEach(model.state.trips, id: \.rideId) { trip in
                Text(SupportLabels.trip(trip)).tag(String?.some(trip.rideId))
            }
        }
        .pickerStyle(.menu)
        .labelsHidden()
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal, MageRideSpacing.sm)
        .frame(minHeight: MageRideControl.minimumTapTarget)
        .background(
            MageRideColor.surfaceVariant,
            in: RoundedRectangle(cornerRadius: MageRideRadius.md, style: .continuous)
        )
    }

    /// *"📎 Attach a screenshot"*, and what it says once one is attached.
    ///
    /// Drawn as an ``OutlinedAction`` rather than being one: `PhotosPicker`'s label is its own
    /// trigger, so a button wrapping it would be a control inside a control.
    private var attachButton: some View {
        PhotosPicker(selection: $picked, matching: .images, photoLibrary: .shared()) {
            HStack(spacing: MageRideSpacing.xxs) {
                Image(systemName: model.state.screenshotName == nil ? "paperclip" : "checkmark.circle.fill")
                    .font(.footnote)
                Text(key: model.state.screenshotName == nil ? "support_attach" : "support_attached")
                    .mageFont(.bodySmall)
            }
            .foregroundStyle(model.state.screenshotName == nil ? MageRideColor.primary : MageRideColor.success)
            .frame(maxWidth: .infinity, minHeight: MageRideControl.outlinedAction)
            .background(
                MageRideColor.background,
                in: RoundedRectangle(cornerRadius: MageRideRadius.md, style: .continuous)
            )
            .overlay {
                RoundedRectangle(cornerRadius: MageRideRadius.md, style: .continuous)
                    .strokeBorder(MageRideColor.outline, lineWidth: MageRideControl.hairline * 2)
            }
            .contentShape(Rectangle())
        }
    }

    /// Reads the picked image into memory.
    ///
    /// **Bytes, not the picker's item.** A `PhotosPickerItem` is valid for the session that produced
    /// it and this one has to survive until Submit — the same reason `apps/passenger-android`'s
    /// `PickedScreenshot` reads the `Uri` at the pick rather than parking it in state. The name is
    /// the app's own and not the file's: `IObjectStore`'s key is built from ids the platform minted,
    /// never from a client filename (backend/CLAUDE.md).
    private func load(_ item: PhotosPickerItem?) async {
        guard let item, let data = try? await item.loadTransferable(type: Data.self) else { return }
        model.onScreenshotPicked(fileName: Self.screenshotFileName, data: data)
    }

    private static let screenshotFileName = "support-screenshot.jpg"
}

/// One ticket and its whole conversation (US-16.2).
@MainActor
private struct TicketThreadSheet: View {

    let ticket: TicketDetail?

    var body: some View {
        ScrollView {
            Group {
                if let ticket {
                    thread(of: ticket)
                } else {
                    LoadingRow(messageKey: "support_loading")
                }
            }
            .padding(MageRideSpacing.md)
        }
        .background(MageRideColor.background)
        .presentationDetents([.height(MageRideControl.ticketSheetHeight), .large])
    }

    private func thread(of ticket: TicketDetail) -> some View {
        VStack(alignment: .leading, spacing: MageRideSpacing.sm) {
            HStack(spacing: MageRideSpacing.xs) {
                Text(SupportLabels.category(ticket.category))
                    .mageFont(.title)
                    .foregroundStyle(MageRideColor.onSurface)
                Spacer(minLength: MageRideSpacing.xs)
                StatusPill(titleKey: SupportLabels.statusKey(ticket.status), tone: SupportLabels.tone(ticket.status))
            }

            // `IosTicketKt.ticketDescription`, not `ticket.description`: that name is `NSObject`'s on
            // this bridge and the Kotlin one is mangled around it — see the helper's own note. It
            // was added by C093 for `apps/driver-ios` and is `:shared`'s, so this app reuses it
            // rather than adding a second `iosMain` function with the same body.
            Text(IosTicketKt.ticketDescription(ticket: ticket))
                .mageFont(.body)
                .foregroundStyle(MageRideColor.onSurface)

            // Oldest first, as the contract sends it: a thread read bottom-up is a thread nobody
            // reads. `assigned` entries are skipped — who is handling a complaint is not the
            // complainant's to see, which is the contract's own rule rather than this screen's.
            //
            // Keyed on the position, because a `TicketEvent` carries no id and two `responded`
            // entries a minute apart are genuinely two rows. Indices rather than
            // `Array(_:enumerated())`: **Swift has no key paths into tuples** (the C087 finding).
            ForEach(ticket.thread.indices, id: \.self) { index in
                threadEntry(ticket.thread[index])
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    /// One `TicketEvent` — what happened, and the agent's words when there are any.
    @ViewBuilder
    private func threadEntry(_ event: TicketEvent) -> some View {
        if let labelKey = SupportLabels.eventKey(event.kind) {
            VStack(alignment: .leading, spacing: 1) {
                Text(key: labelKey)
                    .mageFont(.caption)
                    .foregroundStyle(MageRideColor.onSurfaceVariant)
                if let body = event.body, !body.isEmpty {
                    Text(body)
                        .mageFont(.body)
                        .foregroundStyle(MageRideColor.onSurface)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
        }
    }
}
